### PR TITLE
Add tooltip support to entity editor

### PR DESCRIPTION
**Pull request by yusufkinatas**
## 🌟 What is the purpose of this PR?

This PR adds tooltip support on glide-grid cells. This feature will be used for UX purposes like  indicating which property is required in the table, or showing the full text in a cell, if the text is overflowing. 

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1203007126736607/1203191162200434/f) _(internal)_
- [Design](https://www.figma.com/file/gydVGka9FjNEg9E2STwhi2/HASH-App?node-id=4968%3A243891) _(internal)_

## 🔍 What does this change?

- Adds `customGridIcons`, so we can render those icons via `SpriteRenderer` of `glide-data-grid`
- Refactor `table sorting` codes based on new types
- Create `useGridTooltip` hook, and use it with `properties-table`
- Create custom `value-cell` to handle property values
  - This handles only text values for now

## Known Issues

See next-steps below

## 🐾 Next steps

- Property value tooltips for error state
- Property value tooltips for label property
- Make tooltips available on mobile devices
- Prevent tooltip popping-up while user is editing the cell
- Improve tooltip-hiding logic

## ❓ How to test this?

1.  Checkout the branch / view the deployment
1.  Login
2. Copy your entity-id you get from `http://localhost:3000/entity-editor`
3. Open `http://localhost:3000/@{shortname}/entities/{userId}`  using the entity-id you copied on previous step
1.  Confirm that tooltips are working as expected

## 📹 Demo

Screen Shot 2022-10-24 at 18 11 33

https://user-images.githubusercontent.com/39553853/197561062-511aedf1-f5bb-4261-b91f-81a5a7d7227c.mp4

### CODE REVIEW

File: packages/hash/frontend/src/components/GlideGlid/glide-grid.tsx [hunk 1]
```diff
@@ -39,6 +50,14 @@ export const GlideGrid = (props: DataEditorProps) => {
...
+       * icons defined via `headerIcons` are avaiable to be drawn using
+       * glide-grid's `spriteManager.drawSprite`,
+       * which will be used to draw svg icons inside custom cells
+       */
+      headerIcons={customGridIcons}
     />
   );
 };
+
+export const GlideGrid = forwardRef(_GlideGrid);
```
Comment (nathggns) [inline packages/hash/frontend/src/components/GlideGlid/glide-grid.tsx:63]: This will lead to the component being called `_GlideGrid` in dev tools. The right approach is either:

(this is confirmed to work)

```
const GlideGrid = (…) => { … };

const GlideGridForwardRef = forwardRef(GlideGrid);

export { GlideGridForwardRef as GlideGrid };
```

or this, which supposedly works but I haven't verified

```
export const GlideGrid = forwardRef((...) => { … }); 
```

File: packages/hash/frontend/src/components/GlideGlid/use-grid-tooltip.tsx [hunk 2]
```diff
@@ -0,0 +1,111 @@
+import { DataEditorRef } from "@glideapps/glide-data-grid";
+import { Popover } from "@hashintel/hash-design-system";
+import { PopoverPosition, Typography } from "@mui/material";
+import _ from "lodash";
```
Comment (nathggns) [inline packages/hash/frontend/src/components/GlideGlid/use-grid-tooltip.tsx:4]: It's probably better to import what you specifically use

File: packages/hash/frontend/src/components/GlideGlid/use-grid-tooltip.tsx [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+export const useGridTooltip = (
+  gridRef: RefObject<DataEditorRef>,
+): UseGridTooltipResponse => {
+  const popupState = usePopupState({
+    variant: "popover",
+    popupId: "grid-tooltip",
+  });
+
+  // prevents tooltip to stay at cursor position while user scrolls vertically
```
Comment (nathggns) [inline packages/hash/frontend/src/components/GlideGlid/use-grid-tooltip.tsx:23]: I'm not sure this makes sense grammatically – could you rephrase?

File: packages/hash/frontend/src/components/GlideGlid/use-grid-tooltip.tsx [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+        }
+
+        return { left, top };
+      });
+    },
+    [popupState, gridTooltip, gridRef],
+  );
+
+  const hideTooltip = useCallback<TooltipCellProps["hideTooltip"]>(
+    (_col, _row) => {
```
Comment (nathggns) [inline packages/hash/frontend/src/components/GlideGlid/use-grid-tooltip.tsx:60]: Prefixing with underscore usually means unused but these are both used – why are they prefixed?

File: packages/hash/frontend/src/components/GlideGlid/use-grid-tooltip/grid-tooltip-manager.ts [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+   * we'll use the gradient solution below as
+   * replacement of ellipsis (...) on our custom cells
+   * even if a cell has tooltips, or not
+   */
+  private drawBackground() {
+    const { size, gap, args, cellMargin, bgColor } = this;
+    const { ctx, cell, rect } = args;
+    const {
+      data: { tooltips },
+    } = cell;
```
Comment (nathggns) [inline packages/hash/frontend/src/components/GlideGlid/use-grid-tooltip/grid-tooltip-manager.ts:36]:
```suggestion
    const { tooltips } = cell.data;
```

File: packages/hash/frontend/src/components/GlideGlid/use-grid-tooltip/grid-tooltip-manager.ts [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+
+    ctx.fillRect(grdLeft, rect.y, grdWidth, rect.height);
+  }
+
+  private drawTooltips() {
+    const { size, gap, args, cellMargin } = this;
+    const { ctx, cell, rect, col, row, hoverX = -100, theme } = args;
+    const {
+      data: { hideTooltip, showTooltip, tooltips },
+    } = cell;
```
Comment (nathggns) [inline packages/hash/frontend/src/components/GlideGlid/use-grid-tooltip/grid-tooltip-manager.ts:63]:
```suggestion
    const { hideTooltip, showTooltip, tooltips } = cell.data;
```

File: packages/hash/frontend/src/components/GlideGlid/use-grid-tooltip/grid-tooltip-manager.ts (hunk 5 above)
Comment (nathggns) [inline packages/hash/frontend/src/components/GlideGlid/use-grid-tooltip/grid-tooltip-manager.ts:36]:
```suggestion
    const tooltips = cell.data;
```

File: packages/hash/frontend/src/components/GlideGlid/custom-grid-icons.ts [hunk 7]
```diff
@@ -0,0 +1,14 @@
+export enum CustomGridIcon {
+  LABEL = "bpLabel",
+  ERROR = "bpError",
+  ASTERIKS = "bpAsteriks",
```
Comment (nathggns) [inline packages/hash/frontend/src/components/GlideGlid/custom-grid-icons.ts:4]: This is spelt asterisk

File: packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/entity-editor/properties-section/property-table/cells/value-cell.tsx [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+export const renderValueCell: CustomRenderer<ValueCell> = {
+  kind: GridCellKind.Custom,
+  isMatch: (cell: CustomCell): cell is ValueCell =>
+    (cell.data as any).kind === "value-cell",
+  draw: (args, cell) => {
+    const { ctx, rect, theme } = args;
+    const { x, y, height } = rect;
+    const {
+      data: { property },
+    } = cell;
```
Comment (nathggns) [inline packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/entity-editor/properties-section/property-table/cells/value-cell.tsx:26]:
```suggestion
    const { property } = cell.data;
```

File: packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/entity-editor/properties-section/property-table/cells/value-cell/value-cell-editor.tsx [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+  onChange,
+}) => {
+  return (
+    <TextField
+      sx={{ my: "1px" }}
+      autoFocus
+      value={value.data.property.value}
+      onChange={(event) => {
+        const newValue = { ...value };
+        newValue.data.property.value = event.target.value;
```
Comment (nathggns) [inline packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/entity-editor/properties-section/property-table/cells/value-cell/value-cell-editor.tsx:20]: This mutates `value`, which you appear to have attempted to avoid

**Review by nathggns — DISMISSED** [review]

File: packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/entity-editor/properties-section/property-table/cells/value-cell/value-cell-editor.tsx (hunk 9 above)
Comment (yusufkinatas) [reply packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/entity-editor/properties-section/property-table/cells/value-cell/value-cell-editor.tsx:20]: Good catch, used `lodash.cloneDeep` instead

File: packages/hash/frontend/src/components/GlideGlid/custom-grid-icons.ts (hunk 7 above)
Comment (yusufkinatas) [reply packages/hash/frontend/src/components/GlideGlid/custom-grid-icons.ts:4]: 🤦‍♂️

File: packages/hash/frontend/src/components/GlideGlid/use-grid-tooltip.tsx (hunk 4 above)
Comment (yusufkinatas) [reply packages/hash/frontend/src/components/GlideGlid/use-grid-tooltip.tsx:60]: Forgot to remove them after some refactoring, thanks for catching!

File: packages/hash/frontend/src/components/GlideGlid/use-grid-tooltip.tsx (hunk 3 above)
Comment (yusufkinatas) [reply packages/hash/frontend/src/components/GlideGlid/use-grid-tooltip.tsx:23]: Rephrased as `prevent tooltip from staying at the same position when user scrolls vertically`

File: packages/hash/frontend/src/components/GlideGlid/glide-grid.tsx (hunk 1 above)
Comment (yusufkinatas) [reply packages/hash/frontend/src/components/GlideGlid/glide-grid.tsx:63]: Is this an actual issue to worry about? I feel like having `_GlideGrid` is also fine. We can still search "GlideGrid" to see the component in dev tools. Do we really want to add that extra line for removing the underscore from devtools? I didn't get the motivation here.


Screen Shot 2022-10-26 at 10 26 17

**Comment by yusufkinatas** [comment]
Thanks for the review @nathggns, updated all except first comment. Let me know what do you think.

File: packages/hash/frontend/src/components/GlideGlid/use-grid-tooltip/grid-tooltip-manager.ts [hunk 10]
```diff
@@ -0,0 +1,119 @@
+import type { DrawArgs } from "@glideapps/glide-data-grid/dist/ts/data-grid/cells/cell-types";
+import { CustomGridIcon } from "../custom-grid-icons";
+import { TooltipCell } from "./types";
+
+export class GridTooltipManager {
+  // tooltip size
+  private size = 20;
+
+  // gap between tooltips
+  private gap = 10;
```
Comment (CiaranMn) [inline packages/hash/frontend/src/components/GlideGlid/use-grid-tooltip/grid-tooltip-manager.ts:10]: Doesn't block merging this PR, but if you think that these variable names alone are insufficient to understand their purpose, you could call them `tooltipSize` and `gapBetweenTooltips` and lose the comments – and the variables would be self-describing everywhere they appeared

**Review by CiaranMn — APPROVED** [review]
I'm satisfied that you've addressed Nate's feedback, so I'm approving in his absence. Code looks good from a quick skim (one comment about a general principle rather than any changes required)

File: packages/hash/frontend/src/components/GlideGlid/use-grid-tooltip/grid-tooltip-manager.ts (hunk 10 above)
Comment (yusufkinatas) [reply packages/hash/frontend/src/components/GlideGlid/use-grid-tooltip/grid-tooltip-manager.ts:10]: I agree with this, makes sense. Gonna do this on follow-up pr to prevent conflicts.
